execute/internal/cache/commit_report_cache.go

```diff
@@ -0,0 +1,190 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+
+	readerpkg "github.com/smartcontractkit/chainlink-ccip/pkg/reader"
+	"github.com/smartcontractkit/chainlink-ccip/pkg/types/ccipocr3"
+)
+
+// CommitReportsCache optimizes calls to CommitReportsGTETimestamp by caching reports
+// and only fetching new reports since the last successful query.
+type CommitReportsCache struct {
+	lggr               logger.Logger
+	reports            *cache.Cache // key string -> CommitPluginReportWithMeta
+	lastQueryTimestamp time.Time    // timestamp of the most recent query
+}
+
+// NewCommitReportsCache creates a new commit reports cache.
+// Cached items live for messageVisibilityInterval + EvictionGracePeriod, mirroring
+// the lifetime used in commitRootsCache for executed roots.
+func NewCommitReportsCache(lggr logger.Logger, messageVisibilityInterval time.Duration) *CommitReportsCache {
+	ttl := messageVisibilityInterval + EvictionGracePeriod
+	return &CommitReportsCache{
+		lggr:    lggr,
+		reports: cache.New(ttl, CleanupInterval),
+	}
+}
+
+// generateKey creates a unique key for a commit report
+func generateKey(report ccipocr3.CommitPluginReportWithMeta) string {
+	// We'll create a unique key for each report by combining relevant fields
+	// For each blessed merkle root
+	for _, mrc := range report.Report.BlessedMerkleRoots {
+		return fmt.Sprintf("%s_%s", mrc.ChainSel.String(), mrc.MerkleRoot.String())
+	}
+
+	// If no blessed roots (unlikely but possible), check unblessed roots
+	for _, mrc := range report.Report.UnblessedMerkleRoots {
+		return fmt.Sprintf("%s_%s", mrc.ChainSel.String(), mrc.MerkleRoot.String())
+	}
+
+	// Fallback if no merkle roots (should never happen)
+	return fmt.Sprintf("%d_%d", report.Timestamp.Unix(), report.BlockNum)
```

Comment:
Why would this never happen? I don't see any code filtering out reports that don't have any roots (blessed or otherwise). IMO there's no benefit to wasting memory and storing these in the cache.